theano/tensor/nnet/abstract_conv.py

```diff
@@ -177,6 +179,383 @@ def conv2d(input,
+    for ii in range(-ratio + 1, ratio):
+        kern[ii + ratio - 1] = ratio - abs(ii)
+    if normalize:
+        kern /= ratio
+    return kern
+
+
+def bilinear_upsampling(input,
+                        ratio,
+                        input_shape=None,
+                        use_1D_kernel=True,
+                        repeat_first_row=True,
+                        repeat_first_col=True):
+    """This function takes a stack of 2D inputs and upsamples it (in 2D space)
+    by the ratio. The subsample is assumed to be one and
+
+    :type input: symbolic 4D tensor
+    :param input: mini-batch of feature map stacks, of shape
+        (batch size, input channels, input rows, input columns).
+        See the optional parameter ``input_shape`` that will be upsampled.
+
+    :type ratio: integer
+    : param ratio: the ratio by which the input is upsampled
+        in the 2D space.
+
+    :type input_shape: None, tuple/list of len 4 of int or
+        Constant variable.
+    :param input_shape: The shape of the input variable.
+        Optional, possibly used to choose an optimal implementation.
+        You can give ``None`` for any element of the list to specify that this
+        element is not known at compile time.
+
+    :type use_1D_kernel: bool
+    :param use_1D_kernel: if set to true, row and column will
+        be upsampled seperately by 1D kernels, otherwise they are
+        upsampled together using a 2D kernel. The final result is the
+        same, only the speed can differ, given factors such as
+        upsampling ratio.
+
+    :type repeat_first_row: bool
+    :param repeat_first_row: when upsampling by an even ratio, setting
+        this variable to true copies the first row (making it appear twice
+        in the final upsampled tensor), on the other hand a false value
+        copies the last row.
+
+    :type repeat_first_col: bool
+    :param repeat_first_col: when upsampling by an even ratio, setting
+        this variable to true copies the first column (making it appear twice
+        in the final upsampled tensor), on the other hand a false value
+        copies the last column.
+
+    returns: the upsampled 4D tensor, upsampled using bilinear
+             interpolation with fixed (not learned) kernels such that
+             in the 2D space, each feature is upsampled by the given ratio.
+    """
+
+    bch, ch, row, col = input.shape
+    up_input = input.reshape(-1, 1, row, col)
+
+    # indicating whether the output dimension
+    # is one dimension less (in the case of even ratio)
+    # or not (in the case of odd ratio)
+    if ratio // 2 == 0:
+        sub = 1
+    else:
+        sub = 0
+
+    if use_1D_kernel:
+        kern = bilinear_kernel_1D(ratio=ratio, normalize=True)
+        kern_size = kern.shape[0]
+        # upsampling rows
+        upsampled_row = conv2d_grad_wrt_inputs(output_grad=up_input,
+                                               filters=kern.reshape(1, 1,
```

Comment:
Using indexing with `newaxis` adds `DimShuffle` operations, which can always be done inplace, and which are easier to analyze during the optimization phase (compared to `Reshape`).